app/ide-desktop/lib/dashboard/src/hooks/searchParamsState.ts

```diff
@@ -0,0 +1,70 @@
+/**
+ * @file
+ *
+ * Search params state hook store a value in the URL search params.
+ */
+import * as React from 'react'
+
+import * as reactRouterDom from 'react-router-dom'
+
+import * as eventCallback from '#/hooks/eventCallback'
+
+import * as safeJsonParse from '#/utilities/safeJsonParse'
+
+import * as lazyMemo from './useLazyMemo'
+
+/**
+ * The return type of the `useSearchParamsState` hook.
+ */
+type SearchParamsStateReturnType<T> = Readonly<
+  [value: T, setValue: (nextValue: React.SetStateAction<T>) => void, clear: () => void]
+>
+
+/**
+ * Hook that synchronize a state in the URL search params. It returns the value, a setter and a clear function.
+ * @param key - The key to store the value in the URL search params.
+ * @param defaultValue - The default value to use if the key is not present in the URL search params.
+ */
+export function useSearchParamsState<T = unknown>(
+  key: string,
+  defaultValue: T | (() => T)
+): SearchParamsStateReturnType<T> {
+  const [searchParams, setSearchParams] = reactRouterDom.useSearchParams()
+
+  const lazyDefaultValueInitializer = lazyMemo.useLazyMemo(defaultValue, [])
+
+  const value = React.useMemo<T>(() => {
+    const maybeValue = searchParams.get(key)
+    const defaultValueFrom = lazyDefaultValueInitializer()
+
+    return maybeValue != null
+      ? safeJsonParse.safeJsonParse(maybeValue, defaultValueFrom, (unknown): unknown is T => true)
```

Comment:
dangerous type predicate?!

we have had issues with schemas changing before (previously it was of `localStorage`) - which caused the entire app to crash on every load.

Reply:
is there a reason the validator function can't be passed in?

Reply:
it's quite easy to do validation without runtime schema validation (like zod) - note that `localStorage` uses fully type-safe validation.

> Using `val => val is SomeType` doesn't make your code safer, it's only for typecasting

yes, this is the exact reason we *don't* want to be doing this. at the very least not something as unsafe as `() => true`

Reply:
![CleanShot 2024-03-11 at 19 13 32@2x](https://github.com/enso-org/enso/assets/61194245/67c24d36-6094-475d-88db-60428f3884e0)

This isn't much safer than just `() => true`:

1) You have to specify the type _and_ the parser separately, none of them is connected(via reference)(or inferred) thus eventually the type and the parser will diverge
1) Having tryParse function doesn't mean it doesn't have bugs(at least this function doesn't check if the key is keyof `inputBindingsModule.DashboardBindingKey`.
1) Even though we have an ability to specify the parser for a serialized data, it doesn't mean that anyone(except you) will use it and there's no guarantee we won't miss this on code review

Reply:
the parser is required to return an object of the correct type. it's true that the key isn't validated, but that was an intentional decision as extra keys (and missing keys) should not break the app.

also consider that:
a) the `tryParse` function is capable of transforming data
b) in most other places it's a lot more type-safe as the types are a lot simpler. i wouldn't be surprised if this one is the most complex parse function.

it's true that this validator is unsafe though... the `Partial` is *very* dangerous

tl;dr:
1. they should currently be connected
2. not a bug
3. fair point... QA should catch it though. not sure if there's a nice way to programmatically enforce that a parser must be declared (even when considering alternative APIs)

Reply:
> not sure if there's a nice way to programmatically enforce that a parser must be declared

TLDR: If we agree that the type declaration is the mirror of the schema, then yes, otherwise no :)

Reply:
> the type declaration is the mirror of the schema

wdym by schema here btw?

Reply:
Schema is a sort of a type declaration that does 2 things: Provide type and checks in runtime that the given data matches the type declaration. Single source of truth

Reply:
well, the type declaration is guaranteed to be the mirror of the schema due to how this approach works.

the issue with this approach is that you can accidentally have the declaration without the runtime schema (a runtime schema without a type declaration would error of course.)

iow: 
> the type declaration is the mirror of the schema

in the current codebase, this is guaranteed to always be true 100% of the time, given that the schema exists in the first place.

Reply:
it's theoretically possible to make `registerKey` return something, but the return value would have to be a dummy value...

... alternatively we could use an api a bit like `searchParamsState` where each key has its own individual getter/setter. not sure about the tradeoffs... having these be global would mean that they can't really be put in a context without knowing every single key ahead of time - which is doable but i'd imagine violates SRP